build/repo.targets

```diff
@@ -1,56 +1,195 @@
+  <Target Name="CloneRepositories">
+    <ItemGroup>
+      <_CloneRepository Include="$(MSBuildProjectFullPath)">
+        <AdditionalProperties>
+          CloneRepository=%(Repository.Identity);
+          CloneUrl=%(Repository.CloneUrl);
+          CloneBranch=%(Repository.Branch);
+          CloneRepositoryCommit=%(Repository.Commit);
+          ShallowClone=true;
+          UseGateBranch=$(UseGateBranch)
+        </AdditionalProperties>
+      </_CloneRepository>
+    </ItemGroup>
+
+    <RemoveDir Directories="$(_CloneRepositoryRoot)" Condition="Exists('$(_CloneRepositoryRoot)')" />
+    <MakeDir Directories="$(_CloneRepositoryRoot)" />
+
+    <MSBuild Projects="@(_CloneRepository)"
+      Targets="_CloneRepository"
+      BuildInParallel="$(BuildInParallel)" />
+  </Target>
+
+  <Target Name="_CloneRepository">
     <PropertyGroup>
-      <_SakeTargets>$(_SakeTargets):ci-build</_SakeTargets>
+      <ShallowClone Condition="'$(ShallowClone)'=='' OR '$(CloneRepositoryCommit)'!=''">false</ShallowClone>
+      <_CloneArguments>git clone --quiet $(CloneUrl)</_CloneArguments>
+      <_CloneArguments Condition="'$(ShallowClone)'=='true'">$(_CloneArguments) --depth 1</_CloneArguments>
     </PropertyGroup>
-    <MSBuild Projects="$(MSBuildProjectFullPath)"
-      Targets="Sake"
-      Properties="SakeTargets=$(_SakeTargets);RepositoryRoot=$(RepositoryRoot)" />
+
+    <Message Text="Cloning $(CloneRepository) ..." Importance="High" />
+    <Exec
+      Command="$(_CloneArguments) --branch $(CloneBranch)-gate"
+      Condition="'$(UseGateBranch)'=='true'"
+      IgnoreExitCode="true"
+      IgnoreStandardErrorWarningFormat="true"
+      WorkingDirectory="$(_CloneRepositoryRoot)">
+      <Output TaskParameter="ExitCode" PropertyName="GateBranchExitCode" />
+    </Exec>
+
+    <Warning Text="Using $(CloneBranch)-gate for $(CloneRepository)"
+      Condition="'$(GateBranchExitCode)'=='0'" />
+
+    <Exec
+      Command="$(_CloneArguments) --branch $(CloneBranch)"
+      Condition="'$(GateBranchExitCode)'!='0'"
+      WorkingDirectory="$(_CloneRepositoryRoot)" />
+
+    <Exec
+      Command="git reset --quiet --hard $(CloneRepositoryCommit)"
+      WorkingDirectory="$(_CloneRepositoryRoot)$(CloneRepository)"
+      Condition="'$(CloneRepositoryCommit)'!=''" />
   </Target>
 
-  <Target Name="CITest" DependsOnTargets="SetBuildNumber">
-  <PropertyGroup>
-      <_SakeTargets>$(_SakeTargets):ci-test</_SakeTargets>
-    </PropertyGroup>
-    <MSBuild Projects="$(MSBuildProjectFullPath)"
-      Targets="Sake"
-      Properties="SakeTargets=$(_SakeTargets);RepositoryRoot=$(RepositoryRoot)" />
+  <Target Name="BuildRepositories"
+     DependsOnTargets="_FindDotNetPath;_GenerateRestoreGraphSpecs;_GenerateBuildGraph;_UpdateNuGetConfig;_CreateRepositoriesListWithCommits">
+
+    <MSBuild
+      Projects="$(MSBuildThisFileDirectory)RepositoryBuild.targets"
+      Targets="BuildRepositories"
+      Properties="
```

Comment:
You can build up a list of properties incrementally. 
https://github.com/aspnet/KoreBuild/blob/dev/build/targets/KoreBuild.SolutionBuild.targets#L148-L151